Add "universal" zfs-release RPMs, fix key import

tonyhutter opened the pull request:
1. Update the spec files to use `$releasever` instead of hard coding the version number (7.9, 8.3, 8.4 ...) in the repo.
This means you can use zfs-release-el-1.rpm for any version of RHEL/AlmaLinux/Centos, and zfs-release-fedora-1.rpm for
any version of Fedora.
https://github.com/openzfs/zfs/issues/10328

2. Only import our RPM key on CentOS 7:
https://github.com/zfsonlinux/zfsonlinux.github.com/pull/63

tonyhutter commented:
Here's the important changes:
```diff
--- old/zfs-release.spec	2022-04-25 11:50:29.000000000 -0700
+++ new/zfs-release.spec	2022-06-22 15:49:10.000000000 -0700
@@ -1,21 +1,31 @@
+%if 0%{?fedora}
+%global osname      fedora
+%endif
+
+%if 0%{?rhel}
+%global osname      el
+%endif
+
 Name:           zfs-release
-Version:        1
-Release:        5%{?dist}
-Summary:        ZFS on Linux Repository Configuration
+Version:        %{osname}
+Release:        1
+Summary:        OpenZFS Repository Configuration
 
 Group:          System Environment/Base
 License:        BSD
 URL:            http://zfsonlinux.org
-Source1:        zfs.repo
+Source0:        zfs-el.repo
+Source1:        zfs-fedora.repo
 Source10:       RPM-GPG-KEY-zfsonlinux
 BuildRoot:      %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 BuildArch:      noarch
 
 %description
-ZFS on Linux repository for Fedora.
+OpenZFS repository for Fedora and RHEL variants.
 
 %prep
-echo "Nothing to prep"
+%setup -c -n %{name} -T
+cp -a %{SOURCE0} %{SOURCE1} .
 
 %build
 echo "Nothing to build"
@@ -34,8 +44,8 @@
     $RPM_BUILD_ROOT%{_sysconfdir}/pki/rpm-gpg
 
 # Yum .repo files
-%{__install} -p -m644 %{SOURCE1} \
-    $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d
+%{__install} -p -m644 zfs-%{osname}.repo \
+    $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d/zfs.repo
 
 %clean
 rm -rf $RPM_BUILD_ROOT
@@ -46,21 +56,29 @@
 %config(noreplace) %{_sysconfdir}/yum.repos.d/*
 
 %post
-#rpm --import %{_sysconfdir}/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
+# We only need to import the key on RHEL 7 and below
+# https://github.com/zfsonlinux/zfsonlinux.github.com/pull/63
+%if 0%{?rhel} && 0%{?rhel} < 8
+rpm --import %{_sysconfdir}/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
+%endif
```
```diff
--- old/zfs.repo	2022-06-02 13:35:26.000000000 -0700
+++ new/zfs-el.repo	2022-06-22 15:49:46.000000000 -0700
@@ -1,45 +1,45 @@
 [zfs]
-name=ZFS on Linux for EL8 - dkms
-baseurl=http://download.zfsonlinux.org/epel/8.6/$basearch/
+name=ZFS on Linux for EL$releasever - dkms
+baseurl=http://download.zfsonlinux.org/epel/$releasever/$basearch/
 enabled=1
 metadata_expire=7d
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
 
 [zfs-kmod]
-name=ZFS on Linux for EL8 - kmod
-baseurl=http://download.zfsonlinux.org/epel/8.6/kmod/$basearch/
+name=ZFS on Linux for EL$releasever - kmod
+baseurl=http://download.zfsonlinux.org/epel/$releasever/kmod/$basearch/
 enabled=0
 metadata_expire=7d
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
 
 [zfs-source]
-name=ZFS on Linux for EL8 - Source
-baseurl=http://download.zfsonlinux.org/epel/8.6/SRPMS/
+name=ZFS on Linux for EL$releasever - Source
+baseurl=http://download.zfsonlinux.org/epel/$releasever/SRPMS/
 enabled=0
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
 
 [zfs-testing]
-name=ZFS on Linux for EL8 - dkms - Testing
-baseurl=http://download.zfsonlinux.org/epel-testing/8.6/$basearch/
+name=ZFS on Linux for EL$releasever - dkms - Testing
+baseurl=http://download.zfsonlinux.org/epel-testing/$releasever/$basearch/
 enabled=0
 metadata_expire=7d
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
 
 [zfs-testing-kmod]
-name=ZFS on Linux for EL8 - kmod - Testing
-baseurl=http://download.zfsonlinux.org/epel-testing/8.6/kmod/$basearch/
+name=ZFS on Linux for EL$releasever - kmod - Testing
+baseurl=http://download.zfsonlinux.org/epel-testing/$releasever/kmod/$basearch/
 enabled=0
 metadata_expire=7d
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
 
 [zfs-testing-source]
-name=ZFS on Linux for EL8 - Testing Source
-baseurl=http://download.zfsonlinux.org/epel-testing/8.6/SRPMS/
+name=ZFS on Linux for EL$releasever - Testing Source
+baseurl=http://download.zfsonlinux.org/epel-testing/$releasever/SRPMS/
 enabled=0
 gpgcheck=1
 gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
```

behlendorf commented:
Setting the `Version` to a string I suspect is going to give yum/dnf fits when trying to determine the newest version.  Just bumping it to version 2 should do the right thing.  I think we also do want to keep the `${dist}` tag to make it clear what distro the packages were built on.

```
Version:        2
Release:        1%{?dist}
```

I believe `$releasever` only includes the major version number and not the minor.  That's fine for Fedora, but it does mean for the EL packages we'll need to create a new `7` and `8` repositories with packages that track the current EL releases.

tonyhutter commented:
Actually, I may just want to change the name:
 ```
Name:           zfs-release-%{osname}
Version:        1
Release:        0
```
I guess this is technically is the 1.0 release of `zfs-release-${osname}`, since it's a new package name.

Unfortunately `${dist}` is going to contain the major version like "fc36" or "el8".  That's why we'll need to use `${osname}` if we want something generic like "fedora" or "el".

You're right that `$releasever` is 7, 8, 9, etc, but you can specify your own `$releasever` string, like
```
yum install --releasever=8.4 zfs
```
...and it will install the EL 8.4 version of zfs, along with any EL 8.4 dependencies.  This works because Almalinux 8 & 9 conveniently uses the `8.x/`, `9.x/` directory structure in their repos like we do.  It does not work for CentOS 7, since they use the full release versions like `7.9.2009/`, but that's kind of a moot point, as the CentOS mirrors don't keep any older versions available (see http://mirror.centos.org/centos/7.8.2003/readme).

tonyhutter commented:
I ended up doing the `Name:           zfs-release-%{osname}` change as well as changing the version to 2-1, since I released I used 2-1 in the spec file changelog.  Here's the new spec:

```
%if 0%{?fedora}
%global osname      fedora
%endif

%if 0%{?rhel}
%global osname      el
%endif

Name:           zfs-release-%{osname}
Version:        2
Release:        1
Summary:        OpenZFS Repository Configuration

Group:          System Environment/Base
License:        BSD
URL:            http://zfsonlinux.org
Source0:        zfs-el.repo
Source1:        zfs-fedora.repo
Source10:       RPM-GPG-KEY-zfsonlinux
BuildRoot:      %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
BuildArch:      noarch

%description
OpenZFS repository for Fedora and RHEL variants.

%prep
%setup -c -n %{name} -T
cp -a %{SOURCE0} %{SOURCE1} .

%build
echo "Nothing to build"

%install
rm -rf $RPM_BUILD_ROOT

# Create dirs
install -d -m755 \
  $RPM_BUILD_ROOT%{_sysconfdir}/pki/rpm-gpg  \
  $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d

# GPG Key
%{__install} -Dp -m644 \
    %{SOURCE10} \
    $RPM_BUILD_ROOT%{_sysconfdir}/pki/rpm-gpg

# Yum .repo files
%{__install} -p -m644 zfs-%{osname}.repo \
    $RPM_BUILD_ROOT%{_sysconfdir}/yum.repos.d/zfs.repo

%clean
rm -rf $RPM_BUILD_ROOT

%files
%defattr(-,root,root,-)
%{_sysconfdir}/pki/rpm-gpg/*
%config(noreplace) %{_sysconfdir}/yum.repos.d/*

%post
# We only need to import the key on RHEL 7 and below
# https://github.com/zfsonlinux/zfsonlinux.github.com/pull/63
%if 0%{?rhel} && 0%{?rhel} < 8
rpm --import %{_sysconfdir}/pki/rpm-gpg/RPM-GPG-KEY-zfsonlinux
%endif

%changelog
* Wed Jun 22 2022 Todd Zullinger <tmz@pobox.com> - 2-1
- Build Fedora and EL packages from the same source
* Wed Jun 22 2022 Tony Hutter <hutter2@llnl.gov> - 2-1
- Switch to a universal zfs-release RPM for all RHEL versions
- Don't import GPG key on RHEL 8 or newer:
  https://github.com/zfsonlinux/zfsonlinux.github.com/pull/63
* Wed Jan 15 2020 Tony Hutter <hutter2@llnl.gov> - 1-8.1
- Default repository updated for 8.1 compatible kmod packages
* Fri Aug 25 2017 Brian Behlendorf <behlendorf1@llnl.gov> - 1-5
- Default repository updated for 7.4 compatible kmod packages
* Mon Dec 12 2016 Brian Behlendorf <behlendorf1@llnl.gov> - 1-4
- Default repository updated for 7.3 compatible kmod packages
* Thu Aug 25 2016 Brian Behlendorf <behlendorf1@llnl.gov> - 1-3
- Changed repository from archive.zfsonlinux.org to download.zfsonlinux.org
- Added kmod repository to zfs.repo
* Thu Aug 7 2014 Brian Behlendorf <behlendorf1@llnl.gov> - 1-2
- Changed '$releasever' to '7' in zfs.repo for RHEL/CentOS compatibility
* Wed Jul 23 2014 Brian Behlendorf <behlendorf1@llnl.gov> - 1-1
- Initial zfs-release package.
```

behlendorf commented:
Sounds good.  Since we're changing the package name you could go back to version 1, but I think it's fine to bump it to 2.  The name change will also mean no existing installs will upgrade to the new release package, but I think that's probably for the best.

> You're right that $releasever is 7, 8, 9, etc, but you can specify your own $releasever string, 

Oh that's neat.  Then we'll just need to get the 7, 8, 9 directories in place.

tonyhutter commented:
> Oh that's neat. Then we'll just need to get the 7, 8, 9 directories in place.

I updated our repos prior to submitting this PR, so we're good:
```
--- epel ---
6/
7/
7-legacy/
7.1/
7.2/
7.3/
7.4/
7.5/
7.6/
7.7/
7.8/
7.9/
8/
8.0/
8.1/
8.2/
8.3/
8.4/
8.5/
8.6/
```
`7/` and `8/` are symlinks to the latest release and `7-legacy/` is the old `7/` directory (which contained old 0.6.x releases, now copied into `7.1/` and `7.2/`)
